packages/core/src/Repo.ts

```diff
@@ -188,24 +164,33 @@ export class Repo<T = any> {
    * @param changes (optional) If we're already given the changes (e.g. in `applyChanges`), we can
    * pass them in so we don't have to recalculate them.
    */
-  async set(documentId: string, doc: A.Doc<T>) {
+  async set(documentId: string, doc: A.Doc<T>, changes?: A.Change[]) {
     this.log('set', documentId, doc)
-    // look up old doc and generate diff
-    const oldDoc = (await this.rebuildDoc(documentId)) || A.init()
-    const changes = A.getChanges(oldDoc, doc)
 
+    // look up old doc and generate diff
+    if (!changes) {
+      const oldDoc = (await this.rebuildDoc(documentId)) || A.init()
+      try {
+        changes = A.getChanges(oldDoc, doc)
+      } catch (error) {
+        this.log({ error, oldDoc, doc })
+        changes = []
+      }
+    }
     // cache the doc
-    this.log('set: caching', documentId, doc)
     this.docCache.set(documentId, doc)
 
-    // append changes to this document's history
-    if (changes.length > 0) await this.appendChangeSet({ documentId, changes })
+    // only if Automerge actually found changes in the new document...
```

Comment:
This is a change in logic from before: We were updating the snapshot and firing the handlers even if the document hadn't actually changed.